api/REST-API.md

```diff
@@ -1,191 +1,596 @@
 
-```javascript
-Event.getEvent({ eventId: 123 });
+### What is REST?
+
+When we're building an API, there are many choices that can be taken by the developer, and sometimes this can lead to a variety that doesn't necessarily align with a consensus. For that matter, REST was introduced to the DEV community in order to have a unified way to build APIs so their usage could become easier.
+
+Again, REST is an acronym. It stands for **RE**presentational **S**tate **T**ransfer, and was first presented by [Roy Fielding in 2000](https://www.ics.uci.edu/~fielding/pubs/dissertation/rest_arch_style.htm). 
+
+Basically, a REST API is simply a style of architecture that we use when we design any type of networked applications. It can be implemented in any language, but for this lesson, we will be using NodeJS.
+
+### Guiding Principles of REST
+
+The guiding principles of REST, or sometimes called constraints, are used to define a RESTful system in its entirety. Basically, they are used to limit the means by which the server will respond to the client's requests. These constraints are believed to bring a better "performance, scalability, simplicity, modifiability, visibility, portability, and reliability." ([Wikipidia](https://en.wikipedia.org/wiki/Representational_state_transfer#cite_note-Fielding-Ch5-3)).
+
+So, if you want to build a "true RESTful API", you will want to adhere to the following constrains:
+
+1. Uniform interface: *Fundamental. Simplifies the architecture and enables consensus.*
+1. Client–server: *Separation between client and server; ensures portability.*
+1. Stateless: *No client context stored on server in between requests.*
+1. Cacheable: *Cache responses; improves scalability and performance.*
+1. Layered system: *Enforcing security policies.*
+1. Code on demand (optional): *Transfer executable code.*
+
+([restfulapi.net](https://restfulapi.net/rest-architectural-constraints/))
+
+This is quite a theoretical aspect and since this lesson is aimed at being more practical, you are invited to continue your reading about the RESTful constrains with the couple of articles listed in the [additional readings](#additional-readings) section.
+
+### HTTP Methods
+
+When you want to make a request to your API, if that one follows RESTful conventions, you will want to use specific HTTP Methods. Here are the most common ones; the ones we will be using below in the guided practice:
+
+|HTTP Method|Description|CRUD mapping|
+|-|-|-|
+|`GET`|Fetch data from a collection of resources|`READ`|
+|`POST`|Submit data to a collection of resources|`EDIT`|
+|`PUT`|Update an entire specific resource in a collection|`UPDATE`|
+|`PATCH`|Update part of a specific resource in a collection|`UPDATE`|
```

Comment:
Do people still use patch? I might suggest taking it out and noting it somewhere else...

Reply:
I also mention it because it’s a little easier to remember 4 things when you teach them CRUD.

Reply:
Yes, I agree.

see [1c62c05](https://github.com/Techtonica/curriculum/pull/1410/commits/1c62c05a5fef0b7f4e4a3eba5ae6459a5019094c)